lib/pbench/server/api/resources/datasets_list.py

```diff
@@ -109,6 +126,164 @@ def get_paginated_obj(
+        and
+            "?filter=a.b:c&filter=^b.c:d&filter:^c.d:e"
+
+        result in identical filtering.
+
+        NOTE on implementation:
+
+        SQLAlchemy objects overload most builtin Python operators; and of
+        particular note here, we can separate comparison expressions and
+        capture them in isolation for combination later. For example, the
+        expression "Metadata.user_id == user_id" is not executed immediately;
+        it constructs SQLAlchemy expression objects which will be essentially
+        compiled into SQL phrases later when the query is executed. Similarly,
+        "Metadata.value[("x", "y")]" binds a reference to a second-level JSON
+        field, which would be `value["x"]["y"]` in Python. So we can build
+        very flexible arrays of AND- and OR- terms and combine them as desired.
+
+        Args:
+            filters: A list of filter expressions
+            query: the incoming SQLAlchemy Query with filters attached
+
+        Returns:
+            An updated query with additional filters attached.
+        """
+
+        # Get the authenticated user ID, if any: we only need this for user
+        # namespace keys, but we might process more than one so get it here.
+        user_id = Auth.get_current_user_id()
+
+        or_list = []
+        and_list = []
+        for kw in filters:
+            combine_or = False
+            contains = False
+            try:
+                k, v = kw.split(":", maxsplit=1)
+            except ValueError:
+                raise APIAbort(
+                    HTTPStatus.BAD_REQUEST, f"filter {kw!r} must have the form 'k=v'"
+                )
+            if k.startswith("^"):
+                combine_or = True
+                k = k[1:]
+            if v.startswith("~"):
+                contains = True
+                v = v[1:]
+
+            if not Metadata.is_key_path(k, Metadata.METADATA_KEYS, metalog_key_ok=True):
+                raise APIAbort(HTTPStatus.BAD_REQUEST, str(MetadataBadKey(k)))
+            keys = k.split(".")
+            native_key = keys.pop(0).lower()
+            terms = []
+            use_dataset = False
+            user_private = None
+
+            if native_key == Metadata.DATASET:
+                second = keys[0].lower()
+                # The dataset namespace requires special handling because
+                # "dataset.metalog" is really a special native key space
+                # named "metalog", while other "dataset" sub-keys are primary
+                # columns in the Dataset table.
+                if second == Metadata.METALOG:
+                    native_key = keys.pop(0).lower()
+                else:
+                    try:
+                        c = getattr(Dataset, second)
+                        if c.type.python_type is str:
+                            column = c
+                        else:
+                            column = cast(getattr(Dataset, second), String)
```

Comment:
Reviewing the coverage from the tests, this line does not appear to be tested.  All the other missing coverage are for error handling, which I would love to see tested, but we really should ensure we have this code path covered.